velox/functions/prestosql/ArraySum.cpp

```diff
@@ -0,0 +1,222 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/ComparatorUtil.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+/// Implements the array_sum function.
+///
+template <typename IT, typename OT>
+class ArraySumFunction : public exec::VectorFunction {
+public:
+ // Execute function.
+ void apply(
+     const SelectivityVector& rows,
+     std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+     const TypePtr& outputType,
+     exec::EvalCtx* context,
+     VectorPtr* result) const override;
+};
+
+template <typename IT, typename OT>
+void ArraySumFunction<IT, OT>::apply(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+    const TypePtr& outputType,
+    exec::EvalCtx* context,
+    VectorPtr* result) const {
+  // Acquire the array elements vector.
+  auto arrayVector = args[0]->as<ArrayVector>();
+  VELOX_CHECK(arrayVector);
+  auto elementsVector = arrayVector->elements();
+
+  auto elementsRows =
+      toElementRows(elementsVector->size(), rows, arrayVector);
+  exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+  vector_size_t numRows = arrayVector->size();
+
+  // Allocate new vector for the result
+  memory::MemoryPool* pool = context->pool();
+  auto resultVector = BaseVector::create(outputType, numRows, pool);
+
+
+  // Get access to raw values for the result
+  OT* resultValues = (OT*) resultVector->valuesAsVoid();
+
+  // Iterate over the input vector and find the sum of each array's values
+  for (int i = 0; i < numRows; i++) {
+    // If the whole array is null then set the row null in the output
+    if (arrayVector->isNullAt(i)) {
+      resultVector->setNull(i, true);
+    }
+    // If the array is not null then sum the elements and set the result to the sum
+    else {
+      int start = arrayVector->offsetAt(i);
+      int end = start + arrayVector->sizeAt(i);
+
+      OT sum = 0;
+      for (; start < end; start++) {
+        if (!elements->isNullAt(start)) {
+          sum += elements->template valueAt<IT>(start);
+        }
+      }
+
+      // Set the value at i equal to the sum
+      resultValues[i] = sum;
+    }
+  }
+
+  context->moveOrCopyResult(resultVector, rows, result);
+}
+
+template <>
+void ArraySumFunction<Timestamp, int64_t>::apply(
```

Comment:
The main apply function will not compile for Timestamp, StringView, and Date. Since these specializations are never used (due to the acceptable signatures of array_sum) I have added the specializations that don't compile as no-op functions which lets things compile.

Reply:
Can't we follow the `applyTyped` + `VELOX_DYNAMIC_SCALAR_TEMPLATE_TYPE_DISPATCH` approach as in `ArrayMinMax.cpp`? That would clean up these?